### ISSUE TITLE
Phase 59.4: QuantizedFineTuner — QLoRA & Mixed-Precision Adaptation

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1117]
### Summary
Implement a **QuantizedFineTuner** that enables fine-tuning of quantized large language models using QLoRA and related techniques. By combining 4-bit NormalFloat (NF4) quantization with LoRA adapters, this module enables fine-tuning of 65B+ parameter models on a single 48GB GPU while maintaining full 16-bit fine-tuning quality.

#### QLoRA Core (Dettmers et al.)
- **4-bit NormalFloat quantization**: NF4 data type optimized for normally-distributed weights
- **Double quantization**: Quantize the quantization constants to save additional memory
- **Paged optimizers**: Use NVIDIA unified memory for gradient checkpointing overflow
- **LoRA on quantized base**: Apply LoRA adapters to 4-bit quantized model weights
- **BF16 compute**: Perform adapter computations in BFloat16 for training stability

#### Quantization Backends
- **bitsandbytes integration**: 4-bit and 8-bit quantization via bitsandbytes library
- **GPTQ support**: Post-training quantization via GPTQ algorithm (Frantar et al.)
- **AWQ support**: Activation-aware weight quantization (Lin et al.)
- **GGML/GGUF**: CPU-friendly quantization formats for edge deployment
- **Auto-quantization**: Select optimal quantization scheme based on hardware

#### Mixed-Precision Training
- **Gradient accumulation**: FP32 gradient accumulation with quantized forward pass
- **Loss scaling**: Dynamic loss scaling to prevent gradient underflow
- **Selective precision**: Critical layers in higher precision, others quantized
- **Memory profiling**: Real-time GPU memory tracking with OOM prevention

#### Memory Optimization
- **Gradient checkpointing**: Trade compute for memory via activation recomputation
- **Flash Attention integration**: Memory-efficient attention for long contexts
- **CPU offloading**: Offload optimizer states and gradients to CPU RAM
- **Model parallelism**: Automatic model sharding across available GPUs
- **Batch size finder**: Auto-detect maximum batch size for available memory

### API Design
```python
class QuantizedFineTuner:
    def quantize_model(self, model, config: QuantConfig) -> QuantizedModel
    def apply_qlora(self, model: QuantizedModel, lora_config: LoRAConfig) -> PeftModel
    def profile_memory(self, model, batch_size: int) -> MemoryProfile
    def find_max_batch_size(self, model, sequence_length: int) -> int
    def train(self, model, dataset, training_args: TrainingConfig) -> TrainingResult
    def export_for_inference(self, model, format: str = "merged") -> ExportedModel

class QuantConfig:
    bits: int = 4  # 4 or 8
    quant_type: str = "nf4"  # nf4, fp4, int8
    double_quant: bool = True
    compute_dtype: str = "bfloat16"
    use_paged_optimizer: bool = True
    backend: str = "bitsandbytes"  # bitsandbytes, gptq, awq
```

### Performance Targets
| Metric | Target |
|--------|--------|
| Memory reduction | > 75% vs FP16 fine-tuning |
| Quality (MMLU) | Within 1% of FP16 fine-tuning |
| Training throughput | > 50% of FP16 speed |
| Max model size (48GB GPU) | 65B+ parameters |
| Quantization overhead | < 5 minutes for 7B model |

### References
- Dettmers et al. (2023) — QLoRA: Efficient Finetuning of Quantized Large Language Models
- Frantar et al. (2023) — GPTQ: Accurate Post-Training Quantization for GPT
- Lin et al. (2024) — AWQ: Activation-aware Weight Quantization
- Dettmers et al. (2022) — LLM.int8(): 8-bit Matrix Multiplication for Transformers
- Dettmers & Zettlemoyer (2023) — The Case for 4-bit Precision

### Depends On
- Phase 59.1 (LoRAEngine) for LoRA adapter application
- Phase 59.5 (PEFTOrchestrator) for unified pipeline
- Phase 45 (Knowledge Distillation & Model Compression) for quantization foundations

### Part Of
- Phase 59 — Parameter-Efficient Fine-Tuning & Adaptation
- Planning: #1113
